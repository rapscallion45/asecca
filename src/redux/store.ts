import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
