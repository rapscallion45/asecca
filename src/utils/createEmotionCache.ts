import createCache from '@emotion/cache';

/* create emotion style cache, and set 'prepend' true to allow MUI style overrides */

const createEmotionCache = () => createCache({ key: 'css', prepend: true });

export default createEmotionCache;
