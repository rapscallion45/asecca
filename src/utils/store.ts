/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
