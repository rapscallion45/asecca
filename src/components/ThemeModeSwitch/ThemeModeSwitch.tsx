import { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FormControlLabel, Switch, Box, useTheme, styled } from '@mui/material';
import LightModeIcon from '@mui/icons-material/LightMode';
import DarkModeIcon from '@mui/icons-material/DarkMode';
import { AppState } from '@/redux/store';
import { setTheme } from '@/redux/slices/themeSlice';

/**
 * Theme Mode Switch Style
 *
 * Application light/dark mode toggle switch styling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.6
 *
 * @component
 * @return {Component} - styled theme mode switch
 */
const ThemeModeSwitchStyle = styled(Switch)(({ theme }) => ({
  width: 42,
  height: 26,
  padding: 0,
  '& .MuiSwitch-switchBase': {
    padding: 0,
    margin: 2,
    transitionDuration: '300ms',
    '&.Mui-checked': {
      transform: 'translateX(16px)',
      color: '#fff',
      '& + .MuiSwitch-track': {
        backgroundColor: theme.palette.secondary.main,
        opacity: 1,
        border: 0,
      },
      '&.Mui-disabled + .MuiSwitch-track': {
        opacity: 0.5,
      },
    },
    '&.Mui-focusVisible .MuiSwitch-thumb': {
      color: '#33cf4d',
      border: '6px solid #fff',
    },
    '&.Mui-disabled .MuiSwitch-thumb': {
      color:
        theme.palette.mode === 'light'
          ? theme.palette.grey[100]
          : theme.palette.grey[600],
    },
    '&.Mui-disabled + .MuiSwitch-track': {
      opacity: theme.palette.mode === 'light' ? 0.7 : 0.3,
    },
  },
  '& .MuiSwitch-thumb': {
    boxSizing: 'border-box',
    width: 22,
    height: 22,
  },
  '& .MuiSwitch-track': {
    borderRadius: 26 / 2,
    backgroundColor: theme.palette.mode === 'light' ? '#E9E9EA' : '#39393D',
    opacity: 1,
    transition: theme.transitions.create(['background-color'], {
      duration: 500,
    }),
  },
}));
/**
 * Theme Mode Switch
 *
 * Styled toggle switch used for switching the global app theme
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.6
 *
 * @component
 * @returns {FC} - theme mode switch functional component
 */
const ThemeModeSwitch: FC = () => {
  const dispatch = useDispatch();
  const theme = useTheme();
  const { type: themeType } = useSelector((state: AppState) => state.theme);

  /**
   * Callback that toggles global application theme
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.6
   *
   * @method
   */
  const toggleTheme = () => {
    dispatch(setTheme(themeType === 'dark' ? 'light' : 'dark'));
  };

  return (
    <Box
      display="flex"
      alignItems="center"
      justifyContent="center"
      sx={{
        backgroundColor:
          theme.palette.mode === 'dark' ? '#272727' : theme.palette.grey[200],
        px: 5,
        py: 2,
        mx: 1,
        mt: 3,
        borderRadius: 1,
      }}
    >
      <LightModeIcon
        color={theme.palette.mode === 'dark' ? 'disabled' : 'secondary'}
      />
      <FormControlLabel
        data-testid="light-dark-btn"
        control={
          <ThemeModeSwitchStyle
            focusVisibleClassName=".Mui-focusVisible"
            disableRipple
            checked={themeType === 'dark'}
            onChange={toggleTheme}
            sx={{ ml: 2.5, mr: -1 }}
            color="secondary"
          />
        }
        label=""
      />
      <DarkModeIcon
        color={theme.palette.mode === 'light' ? 'disabled' : 'secondary'}
      />
    </Box>
  );
};

export default ThemeModeSwitch;
