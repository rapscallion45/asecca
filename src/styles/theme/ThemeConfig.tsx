import { FC, useMemo, ReactNode } from 'react';
import { useSelector } from 'react-redux';
import { CacheProvider, EmotionCache } from '@emotion/react';
import { CssBaseline } from '@mui/material';
import {
  ThemeProvider,
  createTheme,
  Theme,
  ThemeOptions,
} from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';
import { AppState } from '@/redux/store';
import lightThemePalette from './lightTheme/lightThemePalette';
import darkThemePalette from './darkTheme/darkThemePalette';
import typography from './typography';

/**
 * Theme Config Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on creating emotion styling engine cache](https://emotion.sh/docs/introduction)
 *
 * @typedef IThemeConfigProps
 * @prop {EmotionCache} emotionCache - emotion styling engine cache
 * @prop {ReactNode} children - component child node
 */
interface IThemeConfigProps {
  emotionCache: EmotionCache;
  children?: ReactNode;
}

/**
 * Theme Config
 *
 * Global application theme definitions and setup
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @component
 * @param {IThemeConfigProps} props - component props
 * @returns {FC} - theme config functional component
 */
const ThemeConfig: FC<IThemeConfigProps> = (props) => {
  const { emotionCache, children } = props;

  /* grab global theme state */
  const { type: themeType } = useSelector((state: AppState) => state.theme);

  /* check if user has dark mode preference */
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');

  /* get options and create theme for current state */
  const themeOptions: ThemeOptions = useMemo(
    () => ({
      palette:
        themeType === 'light' && !prefersDarkMode
          ? lightThemePalette
          : darkThemePalette,
      typography,
    }),
    [themeType, prefersDarkMode]
  );
  const theme: Theme = createTheme(themeOptions);

  return (
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
};

export default ThemeConfig;
