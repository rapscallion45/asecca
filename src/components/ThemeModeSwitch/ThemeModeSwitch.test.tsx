import React from 'react';
import { Provider } from 'react-redux';
import {
  render,
  screen,
  act,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import ThemeModeSwitch from './ThemeModeSwitch';
import ThemeConfig from '../../styles/theme/ThemeConfig';
import createEmotionCache from '../../utils/createEmotionCache';
import store from '../../redux/store';

/**
 * Theme Mode Switch Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.6
 */
describe('Theme Mode Switch', () => {
  it('Renders correctly', () => {
    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <ThemeModeSwitch />
          </ThemeConfig>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Toggle Switch', () => {
    it('Should render theme change button', async () => {
      /** Arrange */
      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <ThemeModeSwitch />
          </ThemeConfig>
        </Provider>
      );

      /** Assert - light/dark theme button should show */
      expect(screen.getByTestId('light-dark-btn')).toBeInTheDocument();
    });

    it('Should toggle switch when theme changed', async () => {
      /** Arrange */
      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <ThemeModeSwitch />
          </ThemeConfig>
        </Provider>
      );

      /** Assert - light/dark theme button should show */
      expect(screen.getByTestId('light-dark-btn')).toBeInTheDocument();
      expect(screen.getByTestId('light-dark-btn')).toHaveProperty(
        'checked',
        false
      );

      /** Act - click theme mode switch to toggle theme color */
      act(() => {
        fireEvent(
          screen.getByTestId('light-dark-btn'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });

      /** Assert - theme switch should have toggled */
      await waitFor(() => {
        expect(screen.getByTestId('light-dark-btn')).toHaveProperty(
          'checked',
          true
        );
      });
    });
  });
});
