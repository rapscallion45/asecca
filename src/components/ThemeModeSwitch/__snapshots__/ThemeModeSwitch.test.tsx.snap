// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Mode Switch Renders correctly 1`] = `
<div
  className="MuiBox-root css-p8oihn"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-uq5yn7-MuiSvgIcon-root"
    data-testid="LightModeIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0 .39-.39.39-1.03 0-1.41l-1.06-1.06zm1.06-10.96c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
    />
  </svg>
  <label
    className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1vbggne-MuiFormControlLabel-root"
    data-testid="light-dark-btn"
  >
    <span
      className="MuiSwitch-root MuiSwitch-sizeMedium css-sj51q3-MuiSwitch-root"
    >
      <span
        className="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary css-7la7f4-MuiButtonBase-root-MuiSwitch-switchBase"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          checked={false}
          className="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          onChange={[Function]}
          required={false}
          type="checkbox"
        />
        <span
          className="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
      </span>
      <span
        className="MuiSwitch-track css-j4fnnk-MuiSwitch-track"
      />
    </span>
    <span
      className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1lyi4-MuiTypography-root"
    />
  </label>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1m9vo64-MuiSvgIcon-root"
    data-testid="DarkModeIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
    />
  </svg>
</div>
`;
